react_juce/core/ImageView.cpp

```diff
@@ -0,0 +1,228 @@
+
+        if (name == sourceProp)
+        {
+            const juce::String source = value.toString();
+            const juce::URL    sourceURL = source;
+
+            if (isWellFormedURL(sourceURL))
+            {
+                // Web images are downloaded in a separate thread to avoid blocking.
+                if (sourceURL.isProbablyAWebsiteURL(sourceURL.toString(false)))
+                {
+                    downloadImageAsync(source);
+                }
+                else if (sourceURL.isLocalFile())
+                {
+                    auto drawableImg = std::make_unique<juce::DrawableImage>();
+                    drawableImg->setImage(loadImageFromFileURL(sourceURL));
+                    drawable = std::move(drawableImg);
+                }
+                else
+                {
+                    const juce::String errorString = "Unsupported image URL: " + source;
+                    throw std::logic_error(errorString.toStdString());
+                }
+            }
+            else if (source.startsWith("data:image/")) // juce::URL does not currently handle Data URLs
+            {
+                auto drawableImg = std::make_unique<juce::DrawableImage>();
+                drawableImg->setImage(loadImageFromDataURL(source));
+                drawable = std::move(drawableImg);
+            }
+            else // If not a URL treat source prop as inline SVG/Image data
+            {
+                drawable = std::unique_ptr<juce::Drawable>(
+                    juce::Drawable::createFromImageData(
+                        source.toRawUTF8(),
+                        source.getNumBytesAsUTF8()
+                    )
+                );
+            }
+        }
+    }
+
+    //==============================================================================
+    void ImageView::paint(juce::Graphics& g)
+    {
+        View::paint(g);
+
+        if (drawable == nullptr) return;
+
+        const float opacity = props.getWithDefault(opacityProp, 1.0f);
+
+        // Without a specified placement, we just draw the drawable.
+        if (!props.contains(placementProp))
+            return drawable->draw(g, opacity);
+
+        // Otherwise we map placement strings to the appropriate flags
+        const int existingFlags = props[placementProp];
+        const juce::RectanglePlacement placement(existingFlags);
+
+        drawable->drawWithin(g, getLocalBounds().toFloat(), placement, opacity);
+    }
+
+    //==============================================================================
+    void ImageView::parentHierarchyChanged()
+    {
+        if (shouldDownloadImage)
+        {
+            downloadImageAsync(props[sourceProp].toString());
+        }
+    }
+
+    //==============================================================================
+    void ImageView::downloadImageAsync(const juce::String& source)
+    {
+        // Allow only one download at the same time for a given ImageView.
+        if (downloading)
```

Comment:
This seems like probably a premature optimization, and so I think we should probably not do this.  Basically a user might change their image source property quite quickly, and we should not cancel the most recent request if the older one is still in-flight. We'll end up with stale state that way.

Let's go ahead without this idea, and if we find that its' really a performance problem we'll come back and revisit it then.

Reply:
Yeah blocking is probably not the right way. But what you're saying made me think of something else. 
If the source is changing quickly, we can't be sure that the last request will finish after previous ones, so there could still be issues. 
And on another level, I think some caching would be interesting to have to avoid new web requests if the same URL is required several times and the source is toggled a lot. It would require some specification but I think it could be useful.

Reply:
> If the source is changing quickly, we can't be sure that the last request will finish after previous ones, so there could still be issues.

Ah yea great point! I didn't think of that

How about this: https://docs.juce.com/master/classImageCache.html :) And this: https://docs.juce.com/master/structDefaultHashFunctions.html#ab3ad9697d3bbe3a07b64f49eddb259f7

So what we can do here is this... during `setProperty`:
* We construct a hash of the source property string using the method I linked above
* If the existing hash already exists in the ImageCache, we can just set the drawable, call repaint(), and return early
* If not, we run the existing code flow.
* Then when we finish downloading an image from the web, we call `juce::MessageManager::callAsync` but in there we do things slightly differently:
    * Insert the image that we just downloaded into the cache
    * Set the `drawable` by reading an Image from the cache based on the hash of the current source prop
    * Repaint

This way, even if requests come in out of order, we will always make sure that the drawable we draw is the correct one for the current source prop. That way too, we can skip subsequent fetches for the same resource and return early.

What do you think?

Reply:
Sounds great to me! I'll push the changes when I'm done :)